obelisk/python/obelisk_py/robot.py

```diff
@@ -0,0 +1,207 @@
+import ctypes
+import multiprocessing
+from abc import ABC, abstractmethod
+from typing import List
+
+import obelisk_sensor_msgs.msg as osm
+import rclpy
+from rclpy.lifecycle.node import LifecycleState, TransitionCallbackReturn
+
+from obelisk_py.node import ObeliskNode
+from obelisk_py.obelisk_typing import ObeliskControlMsg
+
+
+class ObeliskRobot(ABC, ObeliskNode):
+    """Abstract Obelisk robot node.
+
+    Obelisk robots are representations of the physical robot. They take in Obelisk control messages and can optionally
+    output Obelisk sensor messages. We expect code in this function to communicate with the low-level control interface
+    of a real system.
+    """
+
+    def __init__(self, node_name: str) -> None:
+        """Initialize the Obelisk robot."""
+        super().__init__(node_name)
+        self.declare_parameter("sub_ctrl_setting", rclpy.Parameter.Type.STRING)
+        self.declare_parameter("pub_sensor_settings", rclpy.Parameter.Type.STRING_ARRAY)
+
+    def on_configure(self, state: LifecycleState) -> TransitionCallbackReturn:
+        """Configure the robot."""
+        super().on_configure(state)
+
+        # parsing config strings
+        self.sub_ctrl_setting = self.get_parameter("sub_ctrl_setting").get_parameter_value().string_value
+        self.pub_sensor_settings = self.get_parameter("pub_sensor_settings").get_parameter_value().string_array_value
+
+        # create publishers and subscriber
+        self.subscriber_ctrl = self._create_subscription_from_config_str(self.sub_ctrl_setting, self.apply_control)
+        self.publisher_sensors = []
+        # TODO(ahl): under this current version of the code, we cannot auto-create the publishers because we don't know
+        # the message types in both the python and c++ versions of the code.
+        # for sensor_setting in self.pub_sensor_settings:
+        #     pub_sensor = self._create_publisher_from_config_str(sensor_setting)
+        #     self.publisher_sensors.append(pub_sensor)
+
+        return TransitionCallbackReturn.SUCCESS
+
+    def on_cleanup(self, state: LifecycleState) -> TransitionCallbackReturn:
+        """Clean up the robot."""
+        super().on_cleanup(state)
+
+        # destroy publishers + config strings
+        self.subscriber_ctrl.destroy()
+        for sensor_publisher in self.publisher_sensors:
+            self.destroy_publisher(sensor_publisher)
+
+        del self.subscriber_ctrl
+        del self.publisher_sensors
```

Comment:
in this version of the code, is it ok for you to manage this delete, or should the user do it since they make the objects?

Reply:
I think it's fine for us to delete these, since they're the defaults. Down the line, we should expect the user to manage their own deletions. In fact, it's probably OK if they don't even do this, it's pretty benign.